neutral_generation/smart_convert.py

```diff
@@ -0,0 +1,284 @@
+        # e.g. He'shan is a location, but with this rule it would become They'shan
+        s = re.sub(pattern, replace_map[j][1], s)
+
+    return s
+
+
+def capitalize(word):
+    return word[0].upper() + word[1:]
+
+
+def smart_pronoun_replace(s, p, choices):
+    # generate all choices, then choose best option using LM (one with lowest perplexity)
+    # TODO: currently uses OpenAI GPT, can try other LMs to see how performance changes
+    sentence_scores = dict()
+    for choice in choices:
+        new_sentence = regex_token_replace(s, p, new=choice)
+        if s != new_sentence:
+            new_score = score(new_sentence)
+            sentence_scores[new_sentence] = new_score
+
+    if not sentence_scores:   # source pronoun not found in sentence, meaning there are no choices to choose from
+        return s
+
+    print(sentence_scores)
+
+    return min(sentence_scores, key=sentence_scores.get)    # return sentence with lowest score (perplexity)
+
+
+def score(sentence):
+    tokenize_input = tokenizer.tokenize(sentence)
+    tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
+    loss = model(tensor_input, lm_labels=tensor_input)
+    return math.exp(loss) / len(tokenize_input)     # normalize perplexity by number of tokens
+
+
+def identify_verbs_and_auxiliaries(s):
+    """
+    :param s: input sentence
+    :return: dictionary of verbs and list of auxiliaries (if any) with 'she' or 'he' as their subject
+    """
+    SUBJECT_PRONOUNS = ['she', 'he']
+    doc = nlp(s)
+
+    # identify all verbs
+    verbs = set()
+    # this deals with repeating verbs, e.g. "He sings and sings."
+    #   because verb with same text will have different position (makes them unique)
+    for possible_subject in doc:
+        is_subject = (
+            possible_subject.dep == nsubj and                   # current token is a subject
+            # head of current token is a verb
+            (possible_subject.head.pos == VERB or possible_subject.head.pos == AUX) and
+            possible_subject.text.lower() in SUBJECT_PRONOUNS   # current token is either she / he
+        )
+        if is_subject:
+            verbs.add(possible_subject.head)
+
+    # identify all conjuncts and add them to set of verbs
+    # e.g. he dances and prances --> prances would be a conjunct
+    # TODO: understand all English labels of dependency parse better
+    for possible_conjunct in doc:
+        is_conjunct = (
+            possible_conjunct.dep == conj and           # current token is a conjunct
+            # removing the line below, can have a conjunct with an auxiliary verb, e.g. "He was angry and ran away."
+            # possible_conjunct.head.pos == VERB and      # head of current token is a verb
+            possible_conjunct.head in verbs             # the subject of that verb is she / he
+        )
+        if is_conjunct:
+            verbs.add(possible_conjunct)
+
+    # print('verbs: ', verbs)
```

Comment:
Best to remove debugging lines before committing.

Reply:
Got it. Will do.